File: src/main/java/dk/cphbusiness/alg/basics/Container.java
package dk.cphbusiness.alg.basics;

/**
 * Like a Collection in java.util.Collection just without mutability.
 * @param <T>
 */
public interface Container<T> extends Iterable<T> {
  int getSize();
  default boolean isEmpty() { return getSize() == 0; }
  }

File: src/main/kotlin/dk/cphbusiness/alg/basics/Container.kt
package dk.cphbusiness.alg.basics

interface Selection<T> : Iterable<T> {
  fun isEmpty() = size == 0
  val size: Int
  }

interface Bag<T> : Selection<T> {
  fun add(item: T)
  }

interface Stack<T> : Selection<T> {
  fun push(item: T)
  fun pop(): T
  fun peek(): T
  }

interface Queue<T> : Selection<T> {
  fun enqueue(item: T)
  fun dequeue(): T
  fun peek(): T
  }

interface PriorityQueue<T: Comparable<T>> : Selection<T> {
  fun enqueue(item: T)
  fun dequeue(): T
  fun peek(): T
  }
